src/tenant/tests/test_admin.py

```diff
@@ -255,6 +282,149 @@ def test_cant_change_existing_name(self):
         self.assertFalse(form.is_valid())
 
+    def test_cant_create_if_schema_still_exists(self):
+        # delete test tenant object without dropping schema
+        Tenant.get().delete(force_drop=False)
+        # trying to create new tenant, with name of existing schema
+        form = TenantAdminForm(self.form_data)
+        self.assertFalse(form.is_valid())
+
+
+class TenantAdminViewPermissionsTest(TenantTestCase):
+    """Tests for TenandAdmin views permissions."""
+
+    def setUp(self):
+        # create the public schema
+        self.public_tenant = Tenant(schema_name="public", name="public")
+        with tenant_context(self.public_tenant):
+            self.viewuser = User.objects.create_user(
+                username="viewuser",
+                password="sekret",
+                is_staff=True,
+            )
+            self.adduser = User.objects.create_user(
+                username="adduser",
+                password="sekret",
+                is_staff=True,
+            )
+            self.changeuser = User.objects.create_user(
+                username="changeuser",
+                password="sekret",
+                is_staff=True,
+            )
+            self.deleteuser = User.objects.create_user(
+                username="deleteuser",
+                password="sekret",
+                is_staff=True,
+            )
+            # Setup permissions, for our users who can add, change, and delete.
+            opts = Tenant._meta
+
+            # User who can view Tenants
+            self.viewuser.user_permissions.add(
+                get_perm(Tenant, get_permission_codename("view", opts))
+            )
+            # User who can add Tenants
+            self.adduser.user_permissions.add(
+                get_perm(Tenant, get_permission_codename("add", opts))
+            )
+            # User who can change Tenants
+            self.changeuser.user_permissions.add(
+                get_perm(Tenant, get_permission_codename("change", opts))
+            )
+            # User who can delete Tenants
+            self.deleteuser.user_permissions.add(
+                get_perm(Tenant, get_permission_codename("delete", opts))
+            )
+
+            Tenant.objects.bulk_create([self.public_tenant])
+            self.public_tenant.refresh_from_db()
+            self.public_tenant.domains.create(domain="localhost", is_primary=True)
+
+            # create extra tenant for testing purpose
+            self.extra_tenant = Tenant(
+                schema_name="extra",
+                name="extra"
+            )
+            self.extra_tenant.save()
+            domain = self.extra_tenant.get_primary_domain()
+            domain.domain = "extra.localhost"
+
+        self.client = TenantClient(self.public_tenant)
+
+    @override_settings(ROOT_URLCONF=__name__)
+    def test_delete_view(self):
+        """Delete view should restrict access and actually delete items, but leaves schemas in database."""
```

Comment:
"Delete view should restrict access..."<-- can you elaborate on this? What restrictions are actually tested in this test?

"... but leaves schemas in database." This is critical and should be in its own test

Reply:
Django comes with a built-in permissions system. It provides a way to assign permissions to specific users and groups of users, and it’s used by the Django admin site.
       
The Django admin site uses permissions as follows:

- Access to view objects is limited to users with the “view” or “change” permission for that type of object.
- Access to view the “add” form and add an object is limited to users with the “add” permission for that type of object.
- Access to view the change list, view the “change” form and change an object is limited to users with the “change” permission for that type of object.
- Access to delete an object is limited to users with the “delete” permission for that type of object.

Reply:
relevant commit https://github.com/bytedeck/bytedeck/pull/1414/commits/c6b3fe11a1dcb90c0e3d5f0ed5cbd8f5f10e03a4